firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/chunks.txt

```diff
@@ -0,0 +1,282 @@
+      THEN("The input_overwritten flag is false")
+      THEN("The output method reports ok status")
+      THEN("The output buffer is as expected '\x01\x02\x03\x00' ")
+      THEN("The input method reports ok status for the subsequent non-delimeter bytes")
+      THEN("The input_overwritten flag remains unchanged")
+      THEN("The input method reports output_ready status for the delimeter byte")
+      THEN("the output method reports ok status after the last delimeter byte")
+      THEN("The output buffer is as expected '0x04 0x05 0x00' ")       
+
+    WHEN("input to the chunksplitter is '0x01 0x02 0x03 0x00 0x04 0x05 0x00' and output is called after all the bytes are passed")
+      THEN("The input method reports ok status on the first 3 bytes")
+      THEN("The input method reports output_ready status on the 4th byte") // delimeter byte
+      THEN("The input method reports ok status and sets the input_overwritten flag on the 5th byte")
+      THEN("The input method reports ok status for the subsequent byte and input_overwritten flag remains unchanged")
+      THEN("The output method reports ok status")
+      THEN("The output buffer only contains the last chunk '0x04 0x05 0x00' ")  
+
+/* ChunkMerger
+
+Scenario: Protocols::ChunkMerger behaves correctly
+
+  GIVEN("A ChunkMerger with delimiter equal to 0x00")
+    WHEN("A partially full buffer of size 10 bytes is passed as input to transform method")
+      THEN("The transform method reports ok status")
+      THEN("The ChunkMerger appends a delimeter")
+
+    WHEN("Completely full buffer of size 30 bytes is passed as input to transform method")
+      THEN("The transform method reports out_of_bounds status")
+      THEN("The delimeter is not appended to the buffer")
+
+    WHEN("An input data with it's last byte equal to the delimiter is passed to transform")
+      THEN("The transform method reports ok status")
+      THEN("The ChunkMerger appends a delimeter") 
+      // assertion done on buffer as expected
+
+    WHEN("An input data with it's first byte equal to the delimiter is passed to transform")
+      THEN("The transform method reports ok status")
+      THEN("The ChunkMerger appends a delimeter") 
+      // assertion done on buffer as expected
+
+    WHEN("An input data with multiple delimited chunks is passed as input to transform")
+      THEN("The transform method reports ok status")
+      THEN("The ChunkMerger appends a delimeter") 
+
+  GIVEN("A ChunkMerger with delimeter equal to 0x01")
+    WHEN("A partially full buffer of size 10 bytes is passed as input to transform method")
+      THEN("The transform method reports ok status")
+      THEN("The ChunkMerger appends a delimeter")
+
+    WHEN("Completely full buffer of size 30 bytes is passed as input to transform method")
+      THEN("The transform method reports out_of_bounds status")
+      THEN("The delimeter is not appended to the buffer")
+
```

Comment:
Since ChunkSplitter and ChunkMerger represent inverse operations, I think it's worth also designing at least one round-trip test involving two chunks (i.e. we pass the outputs of two ChunkMerger transforms into the inputs of ChunkSplitter and check the correctness of the outputs of the ChunkSplitter after each chunk). This would be a new scenario.

Reply:
```
std::vector merged1.extend([0x01 0x02 0x03] -ChunkMerger-> [0x01 0x02 0x03 0x00])
std::vector merged1.extend([0x04 0x05 0x06] -ChunkMerger-> [0x04 0x05 0x06 0x00])
std::vector merged1.extend([0x07 0x08 0x09] -ChunkMerger-> [0x07 0x08 0x09 0x00])

std::vector<uint8_t> merged1 == [0x01 0x02 0x03 0x00 0x04 0x05 0x06 0x00 0x07 0x08 0x09 0x00]

merged1 -ChunkSplitter-> std::vector split
    0x01 -> ChunkSplitter
    0x02 -> ChunkSplitter
    0x03 -> ChunkSplitter
    std::vector<ByteVector> split.push_back(0x00 -> ChunkSplitter -> [0x01 0x02 0x03])
    ...

std::vector<ByteVector> split == [[0x01 0x02 0x03] [0x04 0x05 0x06] [0x07 0x08 0x09]]

split -ChunkMerger-> std::vector<ByteVector> merged2
    merged2.extend(split[0] -ChunkMerger-> [0x01 0x02 0x03 0x00])
    merged2.extend(split[1] -ChunkMerger-> [0x04 0x05 0x06 0x00])
    merged2.extend(split[2] -ChunkMerger-> [0x07 0x08 0x09 0x00])

std::vector<uint8_t> merged2 == [0x01 0x02 0x03 0x00 0x04 0x05 0x06 0x00 0x07 0x08 0x09 0x00]
```

Reply:
Added here : https://github.com/pez-globo/pufferfish-software/blob/fb237485b1e668d6bc9fa4b0bc9a2bfce24e84e1/firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/chunks.txt#L274